File: SteamPolynomials/src/SteamRegions.f
!-----------------------------------------------------------------------
!
!> @file SteamRegions.f
!
!> @details
!> This module implements function that are capable of determining if
!> a steam state belongs to a particular region according to the revised
!> release on the IAPWS Industrial Formulation 1997 for the Thermodynamic
!> Properties of Water and Steam, August 2007 (IAPWS-IF97)
!
!> @author Panos Asproulis
!> @date 2014
!> @copyright Panos Asproulis (2014-2016). All Rights Reserved.
!
!-----------------------------------------------------------------------
MODULE SteamRegions

    USE SteamConstants

    IMPLICIT NONE
    PRIVATE
    !
    ! Region detection functions
    !
    PUBLIC in_region1
    PUBLIC in_region2
    PUBLIC in_region2a
    PUBLIC in_region2b
    PUBLIC in_region2c
    PUBLIC in_region3
    PUBLIC in_region3a
    PUBLIC in_region3b
    PUBLIC in_region4
    PUBLIC in_region5

CONTAINS
    !-------------------------------------------------------------------------
    !
    !> This function determines whether the specified steam state resides in
    !> region 1 or not.
    !
    !> @param[in] pressure    The steam pressure [MPa]
    !> @param[in] temperature The steam temperature [K]
    !
    !> @return True if the steam state resides in region 1
    !
    !-------------------------------------------------------------------------
    ELEMENTAL FUNCTION in_region1(pressure, temperature)

        USE SteamRegion4, ONLY: psat => saturation_pressure

        IMPLICIT NONE
        !
        ! Arguments
        !
        REAL(KIND=REAL_HIGH), INTENT(IN) :: pressure, temperature
        LOGICAL :: in_region1
        !
        ! Variables
        !
        LOGICAL :: pflag, tflag
        !
        ! Determine whether the steam state belongs in region 1
        !
        tflag = (temperature >= IAPWS97_TMIN) .AND. &
                (temperature <= REGION_1_TMAX)

        pflag = (pressure >= psat(temperature)) .AND. &
                (pressure <= IAPWS97_PMAX)

        in_region1 = .FALSE.

        IF (pflag .AND. tflag) THEN
            in_region1 = .TRUE.
        END IF

    END FUNCTION in_region1
    !-------------------------------------------------------------------------
    !
    !> This function determines whether the specified steam state resides in
    !> region 2 or not.
    !
    !> @param[in] pressure    The steam pressure [MPa]
    !> @param[in] temperature The steam temperature [K]
    !
    !> @return True if the steam state resides in region 2
    !
    !-------------------------------------------------------------------------
    FUNCTION in_region2(pressure, temperature)

        USE SteamRegion4, ONLY: psat => saturation_pressure
        USE SteamBoundaries, ONLY: pbnd => boundary23_pressure

        IMPLICIT NONE
        !
        ! Arguments
        !
        REAL(KIND=REAL_HIGH), INTENT(IN) :: pressure, temperature
        LOGICAL :: in_region2
        !
        ! Determine whether the steam state belongs in region 2
        !
        in_region2 = .FALSE.

        IF ((temperature >= IAPWS97_TMIN) .AND. &
            (temperature <= REGION_1_TMAX)) THEN

            IF ((pressure > 0.0) .AND. &
                (pressure <= psat(temperature))) THEN
                in_region2 = .TRUE.
            END IF

        ELSE IF ((temperature > REGION_1_TMAX) .AND. &
                 (temperature <= REGION_2_4_T)) THEN

            IF ((pressure > 0.0) .AND. &
                (pressure <= pbnd(temperature))) THEN
                in_region2 = .TRUE.
            END IF

        ELSE IF ((temperature > REGION_2_4_T) .AND. &
                 (temperature <= REGION_2_TMAX)) THEN

            IF ((pressure > 0.0) .AND. &
                (pressure <= IAPWS97_PMAX)) THEN
                in_region2 = .TRUE.
            END IF

        END IF

    END FUNCTION in_region2
    !-------------------------------------------------------------------------
    !
    !> This function determines whether the specified steam state resides in
    !> region 2, sub-region 2a.
    !
    !> @param[in] pressure    The steam pressure [MPa]
    !> @param[in] temperature The steam temperature [K]
    !
    !> @return True if the steam state resides in region 2, sub-region 2a.
    !
    !-------------------------------------------------------------------------
    FUNCTION in_region2a(pressure, temperature)

        IMPLICIT NONE
        !
        ! Arguments
        !
        REAL(KIND=REAL_HIGH), INTENT(IN) :: pressure, temperature
        LOGICAL :: in_region2a
        !
        ! Determine whether the steam state belongs in in region 2, sub-region 2a
        !
        in_region2a = .FALSE.

        IF (pressure <= REGION_2A_2B_P) THEN
            in_region2a = in_region2(pressure, temperature)
        END IF

    END FUNCTION in_region2a
    !-------------------------------------------------------------------------
    !
    !> This function determines whether the specified steam state resides in
    !> region 2, sub-region 2b.
    !
    !> @param[in] pressure    The steam pressure [MPa]
    !> @param[in] temperature The steam temperature [K]
    !
    !> @return True if the steam state resides in region 2, sub-region 2b.
    !
    !-------------------------------------------------------------------------
    FUNCTION in_region2b(pressure, temperature)

        USE SteamBoundaries, ONLY: p2bc => boundary2bc_pressure
        USE SteamRegion2, ONLY: h2 => specific_enthalpy

        IMPLICIT NONE
        !
        ! Arguments
        !
        REAL(KIND=REAL_HIGH), INTENT(IN) :: pressure, temperature
        LOGICAL :: in_region2b
        !
        ! Variables
        !
        REAL(KIND=REAL_HIGH) :: h
        !
        ! Determine whether the steam state belongs in region 2, sub-region 2b
        !
        in_region2b = .FALSE.

        h = h2(pressure, temperature)

        IF (pressure > REGION_2A_2B_P) THEN
            IF (pressure <= p2bc(h)) THEN
                in_region2b = in_region2(pressure, temperature)
            END IF
        END IF

    END FUNCTION in_region2b
    !-------------------------------------------------------------------------
    !
    !> This function determines whether the specified steam state resides in
    !> region 2, sub-region 2c.
    !
    !> @param[in] pressure    The steam pressure [MPa]
    !> @param[in] temperature The steam temperature [K]
    !
    !> @return True if the steam state resides in region 2, sub-region 2c.
    !
    !-------------------------------------------------------------------------
    FUNCTION in_region2c(pressure, temperature)

        USE SteamBoundaries, ONLY: p2bc => boundary2bc_pressure
        USE SteamRegion2, ONLY: h2 => specific_enthalpy

        IMPLICIT NONE
        !
        ! Arguments
        !
        REAL(KIND=REAL_HIGH), INTENT(IN) :: pressure, temperature
        LOGICAL :: in_region2c
        !
        ! Variables
        !
        REAL(KIND=REAL_HIGH) :: h
        !
        ! Determine whether the steam state belongs in region 2, sub-region 2c
        !
        in_region2c = .FALSE.

        h = h2(pressure, temperature)

        IF (pressure > p2bc(h)) THEN
            in_region2c = in_region2(pressure, temperature)
        END IF

    END FUNCTION in_region2c
    !-------------------------------------------------------------------------
    !
    !> This function determines whether the specified steam state resides in
    !> region 3 or not.
    !
    !> @param[in] pressure    The steam pressure [MPa]
    !> @param[in] temperature The steam temperature [K]
    !
    !> @return True if the steam state resides in region 3
    !
    !-------------------------------------------------------------------------
    FUNCTION in_region3(pressure, temperature)

        USE SteamBoundaries

        IMPLICIT NONE
        !
        ! Arguments
        !
        REAL(KIND=REAL_HIGH), INTENT(IN) :: pressure, temperature
        LOGICAL :: in_region3
        !
        ! Variables
        !
        LOGICAL :: pflag, tflag
        !
        ! Determine whether the steam state belongs in region 3
        !
        tflag = (temperature >= REGION_1_TMAX) .AND. &
                (temperature <= boundary23_temperature(pressure))

        pflag = (pressure >= boundary23_pressure(temperature)) .AND. &
                (pressure <= IAPWS97_PMAX)

        in_region3 = .FALSE.

        IF (pflag .AND. tflag) THEN
            in_region3 = .TRUE.
        END IF

    END FUNCTION in_region3
    !-------------------------------------------------------------------------
    !
    !> This function determines whether the specified steam state resides in
    !> region 3a or not.
    !
    !> @param[in] density     The steam density [kg/m3]
    !> @param[in] temperature The steam temperature [K]
    !
    !> @return True if the steam state resides in region 3a
    !
    !-------------------------------------------------------------------------
    FUNCTION in_region3a(density, temperature)

        USE SteamRegion3
        USE SteamBoundaries

        IMPLICIT NONE
        !
        ! Arguments
        !
        REAL(KIND=REAL_HIGH), INTENT(IN) :: density, temperature
        LOGICAL :: in_region3a
        !
        ! Variables
        !
        REAL(KIND=REAL_HIGH) :: h3ab, enthalpy, press
        !
        ! Determine whether the steam state belongs in region 3a
        !
        enthalpy = specific_enthalpy(density, temperature)
        press = pressure(density, temperature)
        h3ab = boundary3ab_enthalpy(press)

        IF (enthalpy < h3ab) THEN
            in_region3a = in_region3(press, temperature)
        ELSE
            in_region3a = .FALSE.
        END IF

    END FUNCTION in_region3a
    !-------------------------------------------------------------------------
    !
    !> This function determines whether the specified steam state resides in
    !> region 3b or not.
    !
    !> @param[in] density     The steam density [kg/m3]
    !> @param[in] temperature The steam temperature [K]
    !
    !> @return True if the steam state resides in region 3b
    !
    !-------------------------------------------------------------------------
    FUNCTION in_region3b(density, temperature)

        USE SteamRegion3
        USE SteamBoundaries
        !
        ! Arguments
        !
        REAL(KIND=REAL_HIGH), INTENT(IN) :: density, temperature
        LOGICAL :: in_region3b
        !
        ! Variables
        !
        REAL(KIND=REAL_HIGH) :: h3ab, enthalpy, press
        !
        ! Determine whether the steam state belongs in region 3a
        !
        enthalpy = specific_enthalpy(density, temperature)
        press = pressure(density, temperature)
        h3ab = boundary3ab_enthalpy(press)

        IF (enthalpy >= h3ab) THEN
            in_region3b = in_region3(press, temperature)
        ELSE
            in_region3b = .FALSE.
        END IF

    END FUNCTION in_region3b
    !-------------------------------------------------------------------------
    !
    !> This function determines whether the specified steam state resides in
    !> region 4 or not. Since this region is a line one cannot determine
    !> exactly whether a steam state lies exactly on this line or not so an
    !> error percentage is required for this function to work.
    !
    !> @param[in] pressure    The steam pressure [MPa]
    !> @param[in] temperature The steam temperature [K]
    !> @param[in] error       The acceptable error [%]
    !
    !> @return True if the steam state resides in region 4
    !
    !-------------------------------------------------------------------------
    ELEMENTAL FUNCTION in_region4(pressure, temperature, error)

        USE SteamRegion4, ONLY: tsat => saturation_temperature, &
            psat => saturation_pressure

        IMPLICIT NONE
        !
        ! Arguments
        !
        REAL(KIND=REAL_HIGH), INTENT(IN) :: pressure, temperature, error
        LOGICAL :: in_region4
        !
        ! Variables
        !
        REAL(KIND=REAL_HIGH) :: pressure_estimate, temperature_estimate
        REAL(KIND=REAL_HIGH) :: t_error, p_error
        !
        ! Determine whether the steam state belongs in region 4
        !
        in_region4 = .FALSE.

        IF ((temperature >= IAPWS97_TMIN) .AND. &
            (temperature <= REGION_4_TMAX)) THEN

            pressure_estimate = psat(temperature)
            temperature_estimate = tsat(pressure)

            p_error = ABS(pressure - pressure_estimate) / pressure
            t_error = ABS(temperature - temperature_estimate) / temperature

            IF ((p_error <= error) .AND. (t_error <= error)) THEN
                in_region4 = .TRUE.
            END IF

        END IF

    END FUNCTION in_region4
    !-------------------------------------------------------------------------
    !
    !> This function determines whether the specified steam state resides in
    !> region 5 or not.
    !
    !> @param[in] pressure    The steam pressure [MPa]
    !> @param[in] temperature The steam temperature [K]
    !
    !> @return True if the steam state resides in region 5
    !
    !-------------------------------------------------------------------------
    ELEMENTAL FUNCTION in_region5(pressure, temperature)

        IMPLICIT NONE
        !
        ! Arguments
        !
        REAL(KIND=REAL_HIGH), INTENT(IN) :: pressure, temperature
        LOGICAL :: in_region5
        !
        ! Variables
        !
        LOGICAL :: pflag, tflag
        !
        ! Determine whether the steam state belongs in region 5
        !
        tflag = (temperature >= REGION_2_TMAX) .AND. &
                (temperature <= IAPWS97_TMAX)

        pflag = (pressure > 0.0) .AND. &
                (pressure <= REGION_5_PMAX)

        in_region5 = .FALSE.

        IF (pflag .AND. tflag) THEN
            in_region5 = .TRUE.
        END IF

    END FUNCTION in_region5

END MODULE SteamRegions

File: src/steam_regions.rs
// -------------------------------------------------------------------------------------------------
//
//  This library implements the computation of the steam polynomials according to the revised
//  release on the IAPWS Industrial Formulation 1997 for the Thermodynamic Properties of Water
//  and Steam, August 2007 (IAPWS-IF97).
//
//  Copyright (c) 2014-2025 by Dr. Panos Asproulis (p.asproulis@icloud.com).
//  All Rights Reserved.
//
// -------------------------------------------------------------------------------------------------

//! Steam Regions Module
//!
//! This module implements functions that are capable of determining if
//! a steam state belongs to a particular region according to the revised
//! release on the IAPWS Industrial Formulation 1997 for the Thermodynamic
//! Properties of Water and Steam, August 2007 (IAPWS-IF97)

use crate::steam_boundaries;
use crate::steam_constants::*;
use crate::steam_region2;
use crate::steam_region3;
use crate::steam_region4;

///
/// This function determines whether the specified steam state resides in
/// region 1 or not.
///
/// - Arguments:
///   - `pressure`: The steam pressure (MPa).
///   - `temperature`: The steam temperature (K).
///
/// - Returns:
///   - True if the steam state resides in region 1.
///
pub fn in_region1(pressure: f64, temperature: f64) -> bool {
    //
    // Determine whether the steam state belongs in region 1
    //
    let tflag = (IAPWS97_TMIN..=REGION_1_TMAX).contains(&temperature);
    let pflag =
        (pressure >= steam_region4::saturation_pressure(temperature)) && (pressure <= IAPWS97_PMAX);

    pflag && tflag
}

///
/// This function determines whether the specified steam state resides in
/// region 2 or not.
///
/// - Arguments:
///   - `pressure`: The steam pressure (MPa).
///   - `temperature`: The steam temperature (K).
///
/// - Returns:
///   - True if the steam state resides in region 2.
///
pub fn in_region2(pressure: f64, temperature: f64) -> bool {
    //
    // Determine whether the steam state belongs in region 2.
    //
    if (IAPWS97_TMIN..=REGION_1_TMAX).contains(&temperature) {
        if (pressure > 0.0) && (pressure <= steam_region4::saturation_pressure(temperature)) {
            return true;
        }
    } else if (temperature > REGION_1_TMAX) && (temperature <= REGION_2_4_T) {
        if (pressure > 0.0) && (pressure <= steam_boundaries::boundary23_pressure(temperature)) {
            return true;
        }
    } else if (temperature > REGION_2_4_T)
        && (temperature <= REGION_2_TMAX)
        && (pressure > 0.0)
        && (pressure <= IAPWS97_PMAX)
    {
        return true;
    }

    false
}

///
/// This function determines whether the specified steam state resides in
/// region 2, sub-region 2a.
///
/// - Arguments:
///   - `pressure` - The steam pressure (MPa).
///   - `temperature` - The steam temperature (K).
///
/// - Returns:
///   - True if the steam state resides in region 2, sub-region 2a.
///
pub fn in_region2a(pressure: f64, temperature: f64) -> bool {
    //
    // Determine whether the steam state belongs in region 2, sub-region 2a.
    //
    if pressure <= REGION_2A_2B_P {
        return in_region2(pressure, temperature);
    }

    false
}

///
/// This function determines whether the specified steam state resides in
/// region 2, sub-region 2b.
///
/// - Arguments:
///   - `pressure` - The steam pressure (MPa).
///   - `temperature` - The steam temperature (K).
///
/// - Returns:
///   - True if the steam state resides in region 2, sub-region 2b.
///
pub fn in_region2b(pressure: f64, temperature: f64) -> bool {
    //
    // Determine whether the steam state belongs in region 2, sub-region 2b.
    //
    let h = steam_region2::specific_enthalpy(pressure, temperature);

    if pressure > REGION_2A_2B_P && pressure <= steam_boundaries::boundary2bc_pressure(h) {
        return in_region2(pressure, temperature);
    }

    false
}

///
/// This function determines whether the specified steam state resides in
/// region 2, sub-region 2c.
///
/// - Arguments:
///   - `pressure`: The steam pressure (MPa).
///   - `temperature`: The steam temperature (K).
///
/// - Returns:
///   - True if the steam state resides in region 2, sub-region 2c.
///
pub fn in_region2c(pressure: f64, temperature: f64) -> bool {
    //
    // Determine whether the steam state belongs in region 2, sub-region 2c.
    //
    let h = steam_region2::specific_enthalpy(pressure, temperature);

    if pressure > steam_boundaries::boundary2bc_pressure(h) {
        return in_region2(pressure, temperature);
    }

    false
}

///
/// This function determines whether the specified steam state resides in
/// region 3 or not.
///
/// - Arguments:
///   - `pressure`: The steam pressure (MPa).
///   - `temperature`: The steam temperature (K).
///
/// - Returns:
///   - True if the steam state resides in region 3.
///
pub fn in_region3(pressure: f64, temperature: f64) -> bool {
    //
    // Determine whether the steam state belongs in region 3.
    //
    let tflag = (temperature >= REGION_1_TMAX)
        && (temperature <= steam_boundaries::boundary23_temperature(pressure));

    let pflag = (pressure >= steam_boundaries::boundary23_pressure(temperature))
        && (pressure <= IAPWS97_PMAX);

    pflag && tflag
}

///
/// This function determines whether the specified steam state resides in
/// region 3a or not.
///
/// - Arguments:
///   - `density`: The steam density (kg/m3).
///   - `temperature`: The steam temperature (K).
///
/// - Returns:
///   - True if the steam state resides in region 3a.
///
pub fn in_region3a(density: f64, temperature: f64) -> bool {
    //
    // Determine whether the steam state belongs in region 3a.
    //
    let enthalpy = steam_region3::specific_enthalpy(density, temperature);
    let press = steam_region3::pressure(density, temperature);
    let h3ab = steam_boundaries::boundary3ab_enthalpy(press);

    if enthalpy < h3ab {
        return in_region3(press, temperature);
    }

    false
}

///
/// This function determines whether the specified steam state resides in
/// region 3b or not.
///
/// - Arguments:
///   - `density`: The steam density (kg/m3).
///   - `temperature`: The steam temperature (K).
///
/// - Returns:
///   - True if the steam state resides in region 3b.
///
pub fn in_region3b(density: f64, temperature: f64) -> bool {
    //
    // Determine whether the steam state belongs in region 3b.
    //
    let enthalpy = steam_region3::specific_enthalpy(density, temperature);
    let press = steam_region3::pressure(density, temperature);
    let h3ab = steam_boundaries::boundary3ab_enthalpy(press);

    if enthalpy >= h3ab {
        return in_region3(press, temperature);
    }

    false
}

///
/// This function determines whether the specified steam state resides in
/// region 4 or not. Since this region is a line one cannot determine
/// exactly whether a steam state lies exactly on this line or not so an
/// error percentage is required for this function to work.
///
/// - Arguments:
///   - `pressure` - The steam pressure (MPa).
///   - `temperature` - The steam temperature (K).
///   - `error` - The acceptable error (%).
///
/// - Returns:
///   - True if the steam state resides in region 4.
///
pub fn in_region4(pressure: f64, temperature: f64, error: f64) -> bool {
    //
    // Determine whether the steam state belongs in region 4.
    //
    if (IAPWS97_TMIN..=REGION_4_TMAX).contains(&temperature) {
        let pressure_estimate = steam_region4::saturation_pressure(temperature);
        let temperature_estimate = steam_region4::saturation_temperature(pressure);

        let p_error = (pressure - pressure_estimate).abs() / pressure;
        let t_error = (temperature - temperature_estimate).abs() / temperature;

        if (p_error <= error) && (t_error <= error) {
            return true;
        }
    }

    false
}

///
/// This function determines whether the specified steam state resides in
/// region 5 or not.
///
/// - Arguments:
///   - `pressure` - The steam pressure (MPa).
///   - `temperature` - The steam temperature (K).
///
/// - Returns:
///   - True if the steam state resides in region 5.
///
pub fn in_region5(pressure: f64, temperature: f64) -> bool {
    //
    // Determine whether the steam state belongs in region 5.
    //
    let tflag = (REGION_2_TMAX..=IAPWS97_TMAX).contains(&temperature);
    let pflag = (pressure > 0.0) && (pressure <= REGION_5_PMAX);

    pflag && tflag
}
